experimental/Schemes/SpaceGerms.jl

```diff
@@ -63,37 +201,120 @@ end
+ 0
+ 0
+ 0
+
+```
+"""
+function point(X::AnySpaceGermClosedPoint)
   return point_coordinates(inverted_set(OO(X)))
 end
 
 ## currently unused case
-function point(X::SpaceGerm{<:Any, <:Any, <:GermAtGeometricPoint})
+function point(X::AnySpaceGermGeometricPoint)
   return prime_ideal(inverted_set(OO(X)))
 end
 
 Oscar.ring(X::AbsSpaceGerm) = OO(X)
 
-function Oscar.ideal(X::AbsSpaceGerm{<:Ring,<:MPolyQuoLocRing})
+function Oscar.ideal(X::AnySpaceGerm)
     return modulus(OO(X))
 end
 
-function Oscar.ideal(X::AbsSpaceGerm{<:Ring,<:MPolyLocRing})
+function Oscar.ideal(X::SpaceGerm{<:Ring,<:MPolyLocRing})
     return ideal(OO(X),[zero(OO(X))])
 end
 
-@attr SpaceGerm function ambient_germ(X::AbsSpaceGerm{<:Ring,<:MPolyQuoLocRing})
+@doc raw"""
+    ambient_germ(X::AbsSpaceGerm)
+
+Returns the ambient germ of a given germ `(X,p)`.
+
+More precisely, let `(X,p)` be given by `Spec U^{-1}(R /I)`, where `R` is a polynomial
+ring, `I` an ideal of it and `U` the complement of the maximal ideal corresponding
+to `p. Then the ambient germ `Spec U^{-1}R` is returned.
+
+# Examples:
+```jldoctest
+julia> R, (x,y,z) = QQ["x", "y", "z"];
+
+julia> I = ideal(R, [(x-1)*(x^2 - y^2 + z^2)]);
+
+julia> X = Spec(R, I);
+
+julia> XS = SpaceGerm(X,[0,0,0]);
+
+julia> ambient_germ(XS)
+Spectrum
+  of localization
+    of multivariate polynomial ring in 3 variables x, y, z
+      over rational field
+    at complement of maximal ideal of point (0, 0, 0)
+
+```
+"""
+@attr SpaceGerm function ambient_germ(X::AnySpaceGerm)
     Y,_ = germ_at_point(localized_ring(OO(X)))
     return Y
 end
 
-@attr SpaceGerm function ambient_germ(X::AbsSpaceGerm{<:Ring,<:MPolyLocRing})
+@attr SpaceGerm function ambient_germ(X::SpaceGerm{<:Ring,<:MPolyLocRing})
     return X
 end
 
-############################################################################################################
+@doc raw"""
+    defining_ring_element(X::HypersurfaceGerm)
+    defining_ring_elements(X::CompleteIntersectionGerm)
+
+Returns the (fixed) defining element(s) of the ideal of `X` in the ring of the ambient germ of `X`. Note that the return value is not an element of a polynomial ring, but of a localization of a polynomial ring the complement of a maximal ideal. (Hence each such element has a numerator and a denominator.)
+
+Caution: This command is not exported and is only provided for convenience in programming.
+# Examples:
+```jldoctest
+julia> R, (x,y,z) = QQ["x", "y", "z"];
+
+julia> I = ideal(R, [(x-1)*(x^2 - y^2 + z^2)]);
+
+julia> X = Spec(R, I);
+
+julia> XS = HypersurfaceGerm(X,[0,0,0])
+Spectrum
+  of localization
+    of quotient
+      of multivariate polynomial ring in 3 variables x, y, z
+        over rational field
+      by ideal(x^3 - x^2 - x*y^2 + x*z^2 + y^2 - z^2)
+    at complement of maximal ideal of point (0, 0, 0)
+
+julia> Oscar.defining_ring_element(XS)
+-x^3 + x^2 + x*y^2 - x*z^2 - y^2 + z^2
+
+```
+"""
+defining_ring_element(X::HypersurfaceGerm) = X.f
```

Comment:
```suggestion
defining_ring_element(X::HypersurfaceGerm) = X.f::elem_type(...?...)
```
This should have type assertion, because it's not a concrete type in the struct. We have to look up the precise type getters.  Same below.

Reply:
I added the element types